File: python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py
from aws_cdk import (
  Stack,
  Duration,
  CfnOutput,
  aws_lambda as _lambda,
  aws_lambda_event_sources as lambda_events,
  aws_s3 as s3,
  aws_s3_notifications as s3n,
  aws_sns as sns,
  aws_sns_subscriptions as sns_subs,
  aws_sqs as sqs,
)
from constructs import Construct


class S3SnsSqsLambdaChainStack(Stack):

  def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
    super().__init__(scope, construct_id)

    lambda_dir = kwargs["lambda_dir"]

    # Note: A dead-letter queue is optional but it helps capture any failed messages
    dlq = sqs.Queue(
      self,
      id="dead_letter_queue_id",
      retention_period=Duration.days(7)
    )
    dead_letter_queue = sqs.DeadLetterQueue(
      max_receive_count=1,
      queue=dlq
    )

    upload_queue = sqs.Queue(
      self,
      id="sample_queue_id",
      visibility_timeout=Duration.seconds(30),
      dead_letter_queue=dead_letter_queue
    )

    sqs_subscription = sns_subs.SqsSubscription(
      upload_queue,
      raw_message_delivery=True
    )

    upload_event_topic = sns.Topic(
      self,
      id="sample_sns_topic_id"
    )

    # This binds the SNS Topic to the SQS Queue
    upload_event_topic.add_subscription(sqs_subscription)

    # Note: Lifecycle Rules are optional but are included here to keep costs
    #       low by cleaning up old files or moving them to lower cost storage options
    s3_bucket = s3.Bucket(
      self,
      id="sample_bucket_id",
      block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
      versioned=True,
      lifecycle_rules=[
        s3.LifecycleRule(
          enabled=True,
          expiration=Duration.days(365),
          transitions=[
            s3.Transition(
              storage_class=s3.StorageClass.INFREQUENT_ACCESS,
              transition_after=Duration.days(30)
            ),
            s3.Transition(
              storage_class=s3.StorageClass.GLACIER,
              transition_after=Duration.days(90)
            ),
          ]
        )
      ]
    )

    # Note: If you don't specify a filter all uploads will trigger an event.
    #       Also, modifying the event type will handle other object operations
    # This binds the S3 bucket to the SNS Topic
    s3_bucket.add_event_notification(
      s3.EventType.OBJECT_CREATED_PUT,
      s3n.SnsDestination(upload_event_topic),
      s3.NotificationKeyFilter(prefix="uploads", suffix=".csv")
    )

    function = _lambda.Function(self, "lambda_function",
                                runtime=_lambda.Runtime.PYTHON_3_9,
                                handler="lambda_function.handler",
                                code=_lambda.Code.from_asset(path=lambda_dir))

    # This binds the lambda to the SQS Queue
    invoke_event_source = lambda_events.SqsEventSource(upload_queue)
    function.add_event_source(invoke_event_source)

    # Examples of CloudFormation outputs
    CfnOutput(
      self,
      "UploadFileToS3Example",
      value="aws s3 cp <local-path-to-file> s3://{}/".format(s3_bucket.bucket_name),
      description="Upload a file to S3 (using AWS CLI) to trigger the SQS chain",
    )
    CfnOutput(
      self,
      "UploadSqsQueueUrl",
      value=upload_queue.queue_url,
      description="Link to the SQS Queue triggered on S3 uploads",
    )
    CfnOutput(
      self,
      "LambdaFunctionName",
      value=function.function_name,
    )
    CfnOutput(
      self,
      "LambdaFunctionLogGroupName",
      value=function.log_group.log_group_name,
    )

File: typescript/s3-sns-lambda-chain/lib/s3-sns-sqs-lambda-chain-stack.ts
import { App, Stack, StackProps, Duration, CfnOutput, RemovalPolicy } from 'aws-cdk-lib';
import { join } from 'path'
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as lambdaNodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as lambdaEvents from 'aws-cdk-lib/aws-lambda-event-sources';
import * as snsSubscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as s3Notifications from 'aws-cdk-lib/aws-s3-notifications';

export default class S3SnsSqsLambdaChainStack extends Stack {
  constructor(scope: App, id: string, props?: StackProps) {
    super(scope, id, props);

    // Create a SQS Queue.
    // A dead-letter queue is optional but it helps capture any failed messages.
    const deadLetterQueue = new sqs.Queue(this, 'CsvUploadDeadLetterQueue', {
      queueName: 'CsvUploadDeadLetterQueue',
      retentionPeriod: Duration.days(7)
    });
    const uploadQueue = new sqs.Queue(this, 'CsvUploadQueue', {
      queueName: 'CsvUploadQueue',
      visibilityTimeout: Duration.seconds(30),
      deadLetterQueue: {
        maxReceiveCount: 1,
        queue: deadLetterQueue
      }
    });

    // Create a SNS Topic.
    const uploadEventTopic = new sns.Topic(this, 'CsvUploadTopic', {
      topicName: 'CsvUploadTopic'
    });

    // Bind the SQS Queue to the SNS Topic.
    const sqsSubscription = new snsSubscriptions.SqsSubscription(uploadQueue, {
      rawMessageDelivery: true
    });
    uploadEventTopic.addSubscription(sqsSubscription);


    // Create an S3 Bucket for uploading files.
    const s3Bucket = new s3.Bucket(this, 'CsvBucket', {
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      //! Change the following in production.
      // This deletes the bucket when the stack is deleted (for easy cleanup).
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    // Binds the S3 bucket to the SNS Topic.
    s3Bucket.addEventNotification(
      // Modify the `s3.EventType.*` to handle other object operations.
      s3.EventType.OBJECT_CREATED_PUT,
      new s3Notifications.SnsDestination(uploadEventTopic), {
      // The trigger will only fire on files with the .csv extension.
      suffix: '.csv'
    });


    // Create a Lambda function that will be triggered by the SQS Queue.
    const lambdaFunction = new lambdaNodejs.NodejsFunction(this, 'CsvUploadEventLambda', {
      functionName: 'CsvUploadEventLambda',
      entry: join(__dirname, '..', 'lambda', 'lambda.ts'),
    });

    // Bind the Lambda to the SQS Queue.
    const invokeEventSource = new lambdaEvents.SqsEventSource(uploadQueue);
    lambdaFunction.addEventSource(invokeEventSource);

    // Outputs to help access the key resources.
    new CfnOutput(this, 'DeadLetterQueueUrl', {
      exportName: 'DeadLetterQueue',
      value: `https://console.aws.amazon.com/sqs/home?region=${this.region}#queue/arn:aws:sqs:${this.region}:${this.account}:${deadLetterQueue.queueName}`
    })
    new CfnOutput(this, 'LambdaLogsUrl', {
      exportName: 'LambdaLogs',
      value: `https://console.aws.amazon.com/cloudwatch/home?region=${this.region}#logsV2:log-groups/log-group/$252Faws$252Flambda$252F${lambdaFunction.functionName}`
    })
    new CfnOutput(this, 'UploadCsvToBucketCommand', {
      exportName: 'UploadCsvToBucketCommand',
      value: `aws s3 cp example.csv s3://${s3Bucket.bucketName}/example.csv --acl bucket-owner-full-control`
    })
  }
}
